batch2/batch/driver.py

```diff
@@ -0,0 +1,470 @@
+            secrets[name] = secret.data if not err else None
+
+        return {
+            'spec': self.spec,
+            'secrets': secrets,
+            'output_directory': self.output_directory
+        }
+
+    async def mark_complete(self, status):
+        self._status = status
+        asyncio.ensure_future(db.pods.update_record(self.name, status=json.dumps(status)))
+
+    def mark_deleted(self):
+        assert not self.deleted
+        self.deleted = True
+        self.remove_from_ready()
+
+    async def unschedule(self):
+        if not self.instance:
+            return
+
+        log.info(f'unscheduling {self.name} cores {self.cores} from {self.instance}')
+        self.instance.unschedule(self)
+        self.instance = None
+        await db.pods.update_record(self.name, instance=None)
+
+    async def schedule(self, inst):
+        async with self.lock:
+            assert not self.instance
+
+            self.remove_from_ready()
+
+            if self.deleted:
+                log.info(f'not scheduling {self.name} on {inst.name}; pod already deleted')
+                return False
+
+            if self._status:
+                log.info(f'not scheduling {self.name} on {inst.name}; pod already complete')
+                return False
+
+            if not inst.active:
+                log.info(f'not scheduling {self.name} on {inst.name}; instance not active')
+                asyncio.ensure_future(self.put_on_ready())
+                return False
+
+            if not inst.healthy:
+                log.info(f'not scheduling {self.name} on {inst.name}; instance not healthy')
+                asyncio.ensure_future(self.put_on_ready())
+                return False
+
+            log.info(f'scheduling {self.name} cores {self.cores} on {inst}')
+
+            inst.schedule(self)
+
+            self.instance = inst
+
+            # FIXME: is there a way to eliminate this blocking the scheduler?
+            await db.pods.update_record(self.name, instance=inst.token)
+            return True
+
+    async def put_on_ready(self):
+        # FIXME: does this need a lock?
+        assert not self.on_ready
+
+        if self._status:
+            log.info(f'{self.name} already complete, ignoring put on ready')
+            return
+
+        if self.deleted:
+            log.info(f'{self.name} already deleted, ignoring put on ready')
+            return
+
+        await self.unschedule()
+
+        await self.driver.ready_queue.put(self)
+        self.on_ready = True
+        self.driver.ready_cores += self.cores
+        self.driver.changed.set()
+
+    def remove_from_ready(self):
+        if self.on_ready:
+            self.on_ready = False
+            self.driver.ready_cores -= self.cores
+
+    async def _request(self, f):
+        try:
+            async with aiohttp.ClientSession(
+                    raise_for_status=True, timeout=aiohttp.ClientTimeout(total=5)) as session:
```

Comment:
I think the general idea is to use the same client session for the lifetime of the application. A pod should probably require a session be provided to it on creation. We can store a session in batch.py's `app`. Batch1 currently does this.